framework/src/interface/packet.rs

```diff
@@ -347,6 +382,46 @@ impl<T: EndOffset, M: Sized + Send> Packet<T, M> {
         }
     }
 
+    /* By Dan Jin: The entire network packet is stored in a mbuf by DPDK. A
+       NetBricks packet is a wrapper with an offset that tells you where a
+       particular header starts in the mbuf.
+
+       If you have a Packet<TcpHeader>, then *self.offset() is where the first
+       byte of the tcp header is*. self.payload_offset() is actually just
+       header.size()---NAMES ARE HORRIBLE. In this example, self.offset() +
+       self.payload_offset() give you where the tcp payload would start.
+
+       Use *insert_header* on a mutable reference to insert a header into an
+       already existing packet, after another header's offset basically. It
+       returns a void Result type, i.e. Err or Ok upon inserting a header int*/
+    #[inline]
+    pub fn insert_header<T2: EndOffset<PreviousHeader = T>>(&mut self, header: &T2) -> Result<()> {
+        unsafe {
+            let packet_len = self.data_len();
+            let header_size = header.offset();
+            let added = (*self.mbuf).add_data_end(header_size);
+
+            let hdr = header as *const T2;
+            let offset = self.offset() + self.payload_offset();
+            if added >= header_size {
+                let fin_dst = self.payload();
+                let dst = if packet_len != offset {
+                    // Need to move down the rest of the data down.
+                    let move_loc = fin_dst.offset(header_size as isize);
+                    let to_move = packet_len - offset;
+                    ptr::copy_nonoverlapping(fin_dst, move_loc, to_move);
+                    fin_dst as *mut T2
+                } else {
+                    fin_dst as *mut T2
+                };
+                ptr::copy_nonoverlapping(hdr, dst, 1);
```

Comment:
the comments/naming helps here. One more question though: why are we doing the `let dst` part? Can't we just lose that whole variable, do the "if we need to move stuff" conditional by itself, and follow it with the `ptr::copy_nonoverlapping(hdr, fin_dst, 1)` ?

Reply:
Needs this separation as the `as` is not casting to that type param, but dst is getting the T2 type param. You'd get a 

```shell
expected type parameter, found u8
```

doing it how you mentioned.

Reply:
sorry, was not being clear. Mostly I was wondering whether we could pull the assignment out of the branch, since both conditions return the same thing.

Reply:
yep. @clinedome got it.